xonsh/prompt/env.py

```diff
@@ -1,39 +1,55 @@
 """Prompt formatter for virtualenv and others"""
 
-import os
+import re
+from pathlib import Path
 
 from xonsh.built_ins import XSH
 
 
-def find_env_name():
-    """Finds the current environment name from $VIRTUAL_ENV or
-    $CONDA_DEFAULT_ENV if that is set.
+def env_name():
+    """Build env_name based on different sources. Respect order of precedence.
+
+    Name from VIRTUAL_ENV_PROMPT will be used as-is.
+    Names from other sources are surrounded with ``{env_prefix}`` and
+    ``{env_postfix}`` fields.
     """
-    env_path = XSH.env.get("VIRTUAL_ENV", "")
-    if env_path:
-        env_name = os.path.basename(env_path)
-    else:
-        env_name = XSH.env.get("CONDA_DEFAULT_ENV", "")
-    return env_name
+    if XSH.env.get("VIRTUAL_ENV_DISABLE_PROMPT"):
+        return
+    virtual_env_prompt = XSH.env.get("VIRTUAL_ENV_PROMPT")
+    if virtual_env_prompt:
+        return virtual_env_prompt
+    virtual_env = XSH.env.get("VIRTUAL_ENV")
+    if virtual_env:
+        venv_path = Path(virtual_env)
+        pyvenv_cfg_prompt = prompt_from_pyvenv_cfg(venv_path)
+        if pyvenv_cfg_prompt:
+            return surround_env_name(pyvenv_cfg_prompt)
+        return surround_env_name(venv_path.name)
+    from_conda = XSH.env.get("CONDA_DEFAULT_ENV")
+    if from_conda:
+        return surround_env_name(from_conda)
+    return
 
 
-def env_name():
-    """Returns the current env_name if it non-empty, surrounded by the
-    ``{env_prefix}`` and ``{env_postfix}`` fields.
+def prompt_from_pyvenv_cfg(venv_path):
+    """Grab the prompt from the venv configuration, if it exists.
+
+    Tries to be resilient to subtle changes in whitespace and quoting in the
+    configuration file format as it adheres to no clear standard.
     """
-    env_name = find_env_name()
-    if XSH.env.get("VIRTUAL_ENV_DISABLE_PROMPT") or not env_name:
-        # env name prompt printing disabled, or no environment; just return
-        return
+    assert isinstance(venv_path, Path), venv_path
+    pyvenv_cfg = venv_path / "pyvenv.cfg"
+    if pyvenv_cfg.is_file():
+        match = re.search(r"prompt\s*=\s*(.*)", pyvenv_cfg.read_text())
+        if match:
```

Comment:
can we cache the compiled regex with lru_cache ? this function is invoked for each new prompt and needs to be as fast as possible

Reply:
I wasn't sure if there is some magic caching going on, that I am not aware of, that's why I didn't do anything yet. If this is the case thoug, the file access is also not so great there.

I have another idea. This might feel kinda dirty, but it would solve all potential speed problems in one fell swoop: when this function is called, this means that VIRTUAL_ENV_PROMPT was not set (as it has highest precedence). So this could be used as the cache.

Reply:
we can't be sure that VIRTUAL_ENV is not updated externally, either by running a source command or such. so we should not be making assumptions on that. Here we can compile the regex part and match, and cache the result from opening and reading a file.

Reply:
ah, ok - good point. What do you think about caching the whole function result based on the path that was passed in?

Reply:
Yes we can cache the whole function as well as the re.compiled pattern to match it. Since it will save some when a new path is hit

Reply:
@obestwalter still the regex is not compiled, please create a function that caches the regex part `r"prompt\s*=\s*(.*)"`. rest of them looks fine.